Add interpreter support for the exception handling proposal

takikawa opened the pull request:
This PR adds interpreter support for the exception handling proposal.

Details about the implementation approach:

  * Try blocks generate metadata tracking the instruction ranges for the handlers and which exception tags are handled (or if a `catch_all` is present). The metadata is stored in a function's `FuncDesc`, and is transferred into the `Frame` when a function call is executed.
  * The stack is unwound when a `throw` is executed. This unwinding also handles tag dispatch to the appropriate catch. The metadata to find the matching handler is looked up in the call `Frame` stack.
  * If a `try-delegate` is present, it is used in the stack unwinding process to skip over to the relevant handler.
  * A separate `exceptions_` stack in call frames tracks caught exceptions that can be accessed via a `rethrow`. The stack is popped on exit from a try block or when exiting via control instructions like `br`.
  * Because stack unwinding relies on finding metadata in the call frame, `return_call` needs to be modified slightly to adjust the current frame when executing the call, rather than re-using the frame completely as-is.

Spec tests will be followed up in future PRs once the proposal tests are merged into the official testsuite.

takikawa commented:
Thanks for the merging the other two PRs @aheejin :) I forgot to also CC you on this big patch.

sbc100 commented:
Can you re-base on top of the already landed parts.  Also, if there are remaining separate commits can you split them out into their own PRs maybe?  Or does that make no sense in this case?

takikawa commented:
> Can you re-base on top of the already landed parts.

Sure, no problem. I just pushed a rebased version. Also I squashed the other commit into this too as it's not really a separate thing (since I was force pushing for the rebase anyway).

sbc100 commented on src/interp/interp.h:
```diff
@@ -305,13 +306,37 @@ struct LocalDesc {
+  u32 offset;
+};
+
+// The Try kind is for catch-less try blocks, which have a HandlerDesc that
```
Is this comment out-of-date?  What is `Try kind` here?

takikawa replied:
Thanks for catching this, yeah it's out-of-date now. I'll more clearly indicate what's included in the `Catch` case.

sbc100 commented on src/interp/istream.h:
```diff
@@ -98,7 +98,7 @@ class Istream {
   // Each opcode is a SerializedOpcode, and each immediate is a u32.
   static const Offset kBrTableEntrySize =
-      sizeof(SerializedOpcode) * 2 + 3 * sizeof(u32);
+      sizeof(SerializedOpcode) * 3 + 4 * sizeof(u32);
```
I'm having trouble understanding how the above comment describes these magic numbers.    Does the comment need updating?  If not, why not?

takikawa replied:
Thanks yes, the part that needs updating is the "made of two instructions" part. It's now three instructions so I'll just add the missing bit to the comment.

sbc100 reviewed (approved):
Nice!

Will some of the tests added here be redundant once we start running the spec tests?

Can you move the notes to the main PR description?

I'll leave the final lgtm to @aheejin

takikawa commented:
Re: tests, there is some redundancy with spec tests in that there's overlap in what's tested and so they look similar (though I didn't copy-paste the spec tests). I did try to write tests that specifically target the more tricky bits of the wabt interpreter implementation (e.g., the exception stack for `rethrow`, how exception handlers are looked up, etc).

sbc100 commented:
lgtm.  WDYT @aheejin ?

takikawa commented:
Friendly ping on this @aheejin, any feedback or comments appreciated :)

aheejin commented:
Sorry! I was OOO for a while. I will take a look shortly.

aheejin reviewed (commented):
Sorry again for the super delayed reply. Haven't read the tests yet, but here are comments and questions. I'm not very familiar with the wabt codebase, so I might be asking some basic questions 😅

takikawa commented on src/type-checker.cc:
```diff
@@ -100,6 +100,24 @@ Result TypeChecker::GetRethrowLabel(Index depth, Label** out_label) {
   return Result::Error;
 }
 
+Result TypeChecker::GetCatchDepth(Index depth, Index* out_depth) {
```
Good point, I made this renaming now.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  // Jump to the end of the block at the end of the previous try or catch.
+  Istream::Offset offset = label->offset;
+  istream_.Emit(Opcode::Br);
+  if (offset == Istream::kInvalidOffset) {
+    depth_fixups_.Append(label_stack_.size() - 1, istream_.end());
+  }
```
I think you're right, I'm not sure why this was an `if`. Might be part of a previous attempt at this patch. I've changed this to an assertion.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -259,9 +280,12 @@ class BinaryReaderInterp : public BinaryReaderNop {
+  void PushLabel(LabelKind label = LabelKind::Block,
+                 Istream::Offset offset = Istream::kInvalidOffset,
+                 Istream::Offset fixup_offset = Istream::kInvalidOffset,
+                 u32 handler_desc_index = 0);
```
Thanks, good point. Applied.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  }
+  istream_.Emit(offset);
+  desc.end_offset = istream_.end();
+  label->kind = LabelKind::Block;
```
Thanks, added a comment.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+    depth_fixups_.Append(label_stack_.size() - 1, istream_.end());
+  }
+  istream_.Emit(offset);
+  desc.end_offset = istream_.end();
```
Thanks, you're right. This is a bug. I fixed it and added test cases for it (`try-catch-nested-4` and `try-catch-nested-5` in `try.txt`---these specifically test throwing an exception in a catch block in a `try-catch` with multiple catches).

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+    depth_fixups_.Append(label_stack_.size() - 1, istream_.end());
+  }
+  istream_.Emit(offset);
+  desc.end_offset = istream_.end();
```
Fixed in previous thread.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  desc.end_offset = istream_.end();
+  Label* target_label = GetNearestTryLabel(depth + 1);
+  assert(target_label);
+  desc.delegate_offset = target_label->handler_desc_index;
```
Thanks, renamed.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  assert(target_label);
+  desc.delegate_offset = target_label->handler_desc_index;
+  FixupTopLabel();
+  PopLabel();
```
So I think most of the checks that `OnEndExpr` are doing are specific to how `end` works, in that it can be paired with different instructions. Also the `Label` it uses is actually a different data structure from the validator. For the stack size check, that's also only for the case in which it's a function body block I think.

I have added an `assert` though that makes sure that the current label is a `LabelKind::Try` instead of possibly a `LabelKind::Block`, which makes this code do more sanity checking that it's only a `try-delegate`. Let me know if that's not enough.

aheejin replied:
Aren't these lines necessary though? https://github.com/WebAssembly/wabt/blob/d61a78988001cdc6d364d91cfb9b2088b9ae2798/src/interp/binary-reader-interp.cc#L1024-L1027

These check 1. if the label is valid and 2. the type and number of the values on the stack and the label's type matches. `try`-`delegate` can act like a block, so it can return values.

takikawa replied:
If I'm understanding correctly, I think these checks are already done by the `OnDelegate` call near L1587. That function's here: https://github.com/WebAssembly/wabt/blob/5d4955c0e819ce0d8943824f3d458894be628fd3/src/type-checker.cc#L572 and it checks the label (`CHECK_RESULT(GetLabel(depth + 1, &label));`) and does `PopAndCheckSignature` and so on.

aheejin replied:
Oh I see.. Thanks.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -837,7 +906,14 @@ Result BinaryReaderInterp::BeginFunctionBody(Index index, Offset size) {
 
   // Push implicit func label (equivalent to return).
-  PushLabel();
+  // With exception handling it acts as a catch-less try block.
```
Good point, I've added this now.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1054,24 +1141,29 @@ Result BinaryReaderInterp::OnBrTableExpr(Index num_targets,
     // instruction has a fixed size.
     istream_.Emit(Opcode::InterpDropKeep, drop_count, keep_count);
-    EmitBr(depth, 0, 0);
+    istream_.Emit(Opcode::InterpCatchDrop, catch_drop_count);
```
Added `"Same for CatchDrop as well"` here.

takikawa commented on src/interp/interp.cc:
```diff
@@ -386,6 +401,11 @@ Result DefinedFunc::DoCall(Thread& thread,
   result = thread.Run(out_trap);
   if (result == RunResult::Trap) {
     return Result::Error;
+  } else if (result == RunResult::Exception) {
```
This is possible, but it's not trivial to add here because we need to track the currently thrown exception (only caught exceptions are stored currently, for rethrow) and then access it here with a new helper function to print the values to a string (`WriteValues` exists, but only operates on streams for tool/test output).

Maybe this could be done in follow-up changes later? WDYT?

aheejin replied:
Yeah that sounds fine. (Also I think doing that is not a high priority)

takikawa commented on src/interp/interp.h:
```diff
@@ -378,13 +403,19 @@ struct ModuleDesc {
+  explicit Frame(Ref func,
+                 u32 values,
+                 u32 offset,
+                 u32 exceptions,
```
Done

takikawa commented on src/interp/interp.cc:
```diff
@@ -2390,6 +2443,87 @@ RunResult Thread::DoAtomicRmwCmpxchg(Instr instr, Trap::Ptr* out_trap) {
   return RunResult::Ok;
 }
 
+RunResult Thread::DoThrow(Ref exn_ref) {
```
I think the purpose of the `Ptr` types is to ensure that your object stays live during interpreter operations, whereas the `Ref` is the representation used for the reftyped value while it lives on the stack, etc. which is trace by the runtime itself.

Yeah as you said it might be more consistent to use a `Ptr` to the exception here. I think in terms of correctness, it should work either way because `DoThrow` immediately creates a `Ptr` to the passed-in `Ref` and ensures that it's live until the end of the function. In addition, an exception is either caught (stored in `exceptions_` and therefore live until popped by control flow out of the catch) or uncaught and doesn't need to be live anymore (this could change if exceptions propagate out to the "host" in a more interesting way in the interpreter).

In any case, I will change this to take a `Ptr` as it's a simple refactoring.

aheejin replied:
I see, thanks. Should we mark `exceptions_` in `Thread::Mark` too? Because apparently `values_` are marked there.

takikawa replied:
Yes, I think you're right. I pushed a small fix for this.

takikawa commented on src/interp/interp.cc:
```diff
@@ -2390,6 +2443,87 @@ RunResult Thread::DoAtomicRmwCmpxchg(Instr instr, Trap::Ptr* out_trap) {
+        // For a try-delegate, skip part of the traversal by directly going
+        // up to an outer handler specified by the delegate depth.
+        if (handler.kind == HandlerKind::Delegate) {
+          iter = handlers.rend() - handler.delegate_offset - 1;
```
In this case, since we're iterating in reverse and want to potentially continue iterating after the delegating (the delegated handler may not handle the tag, etc), we want `iter` to be a reverse iterator. Since the handler index is from the start, we subtract from `rend()` which is equivalent to the start in the other direction (but off by one, because `rend()` is off the end instead of the first element).

I can add a comment to make this clearer.

takikawa commented on src/interp/interp.cc:
```diff
@@ -2390,6 +2443,87 @@ RunResult Thread::DoAtomicRmwCmpxchg(Instr instr, Trap::Ptr* out_trap) {
+        for (auto _catch : handler.catches) {
+          // Here we have to be careful to use the target frame's instance
+          // to look up the tag rather than the throw's instance.
+          Ref catch_ref = frame.inst->tags()[_catch.tag_index];
```
Done

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1100,15 +1192,19 @@ Result BinaryReaderInterp::OnCallIndirectExpr(Index sig_index,
   // will change the type stack.
   CHECK_RESULT(validator_.OnReturnCall(loc, Var(func_index)));
   istream_.EmitDropKeep(drop_count, keep_count);
+  istream_.EmitCatchDrop(catch_drop_count);
```
That's a bug, thanks for catching it. I fixed it and added a test case (`rethrow-return`).

takikawa replied:
BTW, maybe `return_call_indirect` also needs some handling here. I'll check that separately after I push the current patch of changes.

takikawa commented on src/interp/interp.cc:
```diff
@@ -2390,6 +2443,87 @@ RunResult Thread::DoAtomicRmwCmpxchg(Instr instr, Trap::Ptr* out_trap) {
+  // If the throw crosses call frames, we need to reset the state to that
+  // call frame's values.
+  if (popped_frame) {
+    values_.resize(target_frame.values);
+    exceptions_.resize(target_frame.exceptions);
```
You're right, I overlooked this case. I've added some test cases for this (`try-catch-stack-size` and `rethrow-throw`) and fixed it.

The fix is to add the local stack heights for both in the `HandlerDesc`, so that when we look it up here we can resize the vectors to the frame's height plus the local height in the handler. I've added some comments to clarify it too.

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1100,15 +1192,19 @@ Result BinaryReaderInterp::OnCallIndirectExpr(Index sig_index,
+  istream_.EmitCatchDrop(catch_drop_count);
 
   if (func_index >= num_func_imports()) {
+    istream_.Emit(Opcode::InterpAdjustFrameForReturnCall, func_index);
```
Looking at the code, I do notice that `GetLocalType` looks up the function reference from the stack frame, so it might be possible to construct a failing case for other programs too. I'll look into this (will file a separate issue/PR if there's a bug).

takikawa replied:
So it's hard to produce a failing test case for normal operation, the frame behavior of `return_call` did break the tracing in the interpreter. For example:

```
$ wabt/out/clang/Debug/wasm-interp out/test/interp/return-call-get-local/return-call-get-local.wasm -t --run-all-exports --enable-tail-call
>>> running export "f":
#0.    0: V:0  | i64.const 1
#0.   12: V:1  | br @24
#0.   24: V:1  | i32.const 3
#0.   32: V:2  | drop_keep $1 $1
#0.   44: V:1  | br @68
#0.   68: V:1  | i32.const 2
wasm-interp: ../../../src/interp/interp.cc:192: wabt::interp::ValueType wabt::interp::FuncDesc::GetLocalType(wabt::interp::Index) const: Assertion `iter != locals.end()' failed.
Aborted (core dumped)
```

when running the example in issue #1761.

The change in this PR doesn't quite fix this bug (in addition to updating the function reference, the value stack height needs to be updated) but it could be easily done in a follow-up.

aheejin replied:
I didn't mean to ask you to investigate an existing bug, but thanks for the fix!

takikawa commented:
Thanks for the review feedback! I think I've addressed all the feedback now, although I still need to check if `return_call_indirect` has any faulty interaction with this.

takikawa commented:
Pushed an extra commit that should fix the CI failure

takikawa commented:
I pushed a fix for the `return_call_indirect` case in 34fb572. I put in a sanity check assert that checks, when a function call is popped, that the frame's exception stack height (the height on function entry) matches the current height. This assert caught another bug, so I included that fix (some `catch`s weren't popping the exception because it was being done only for `end`).

takikawa commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1497,103 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  }
+  // The label kind is switched to Block from Try in order to distinguish
+  // catch blocks from try blocks. This is needed for operations like
+  // GetNearestTryLabel, where we need to only find try blocks.
```
Ok, I just pushed a rewording for the comment that makes it more explicit about being about delegation and passing catches.

aheejin commented on src/interp/binary-reader-interp.cc:
```diff
@@ -1394,6 +1500,110 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  HandlerDesc& desc = func_->handlers[label->handler_desc_index];
+  desc.kind = HandlerKind::Catch;
+  // Jump to the end of the block at the end of the previous try or catch.
+  if (label->kind == LabelKind::Block) {
+    istream_.EmitCatchDrop(1);
+  }
```
Shoudn't we do this even when there's only a single `catch` or `catch_all`?

takikawa replied:
If there's only a single `catch`, it should get handled by the code path in `OnEnd` here: https://github.com/WebAssembly/wabt/blob/d9cf828520103f916083af2e96aff914800c81f5/src/interp/binary-reader-interp.cc#L1114

(because there's no next catch block that would run this cleanup)

aheejin replied:
```wasm
try
  ...
catch $e0  ;; This gets dropped by this fix
  ...
catch $e1  ;; Where does this get dropped?
  ...
catch $e2  ;; This gets dropped by `OnEnd`
  ...
end
```

takikawa replied:
The second one should get dropped by this fix too, because each `catch` opcode triggers an `OnCatchExpr`call . On the first `OnCatchExpr` it doesn't drop anything as the previous code is for the try body (`label->kind == LabelKind::Try`). On the second, it drops `$e0`. On the third, it drops `$e1`. Finally, `OnEndExpr` drops `$e2`.

aheejin replied:
Sorry again, how about adding one-line comment saying this is dropping the previous catch's exception then?

takikawa replied:
Sure no problem, added.

aheejin commented on test/interp/rethrow-and-br.txt:
```diff
@@ -0,0 +1,125 @@
+                (try
+                  (do (throw $e2))
+                  (catch $e2
+                    ;; exception stack has $e2
```
Doesn't the exception stack have both $e1 and $e2? There's the outer catch in line 90.

takikawa replied:
Yeah that's true, I've now made this comment more precise to indicate what it's really testing.

aheejin commented on test/interp/try.txt:
```diff
@@ -0,0 +1,143 @@
+      (catch $e1)))
+  (func (export "try-catch-stack-size-2") (result i32)
+    (i32.const 1)
+    (call $helper))
+        )
```
```suggestion
    (call $helper)))
```

takikawa replied:
Done.

aheejin reviewed (approved):
Thanks for your patient answers and fixes! Sorry again this took so long. LGTM % remaining small nits

takikawa commented:
Thanks for all the comments! I think I've addressed everything now, and CI looks good.
